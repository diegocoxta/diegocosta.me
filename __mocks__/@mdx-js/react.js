module.exports = {
  MDXProvider: jest.fn(({ children }) => children),
};
