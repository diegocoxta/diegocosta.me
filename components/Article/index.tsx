import { MDXRemote } from 'next-mdx-remote/rsc';

import type { BlogContentAttributes } from '~/app/cms';

import CodeBlock from '~/components/CodeBlock';

import styles from './styles.module.css';

export default function Article(props: { children: BlogContentAttributes['content'] | undefined }): React.ReactElement {
  return (
    <div className={styles.article}>
      {props.children && (
        <MDXRemote
          source={props.children}
          components={{
            code: CodeBlock,
          }}
        />
      )}
    </div>
  );
}
