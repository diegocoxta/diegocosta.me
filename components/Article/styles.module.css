.article {
  line-height: 1.5;
  font-size: 19px;
  margin: 0 auto 60px;
  font-weight: 400;

  :global(.responsive) {
    position: relative;
    padding-bottom: 56.25%;
    height: 0;
    overflow: hidden;
    max-width: 100%;
  }

  :global(.responsive iframe),
  :global(.responsive object),
  :global(.responsive embed) {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
  }

  a {
    color: var(--main-accent-color);
    text-decoration: none;
    border-bottom: 1px solid var(--main-background-color);

    &:hover,
    &:focus {
      border-bottom: 1px solid var(--main-accent-color);
      outline: none;
    }
  }

  iframe {
    width: 100%;
    border: 0;
  }

  blockquote {
    border-left: 5px solid var(--main-accent-color);
    padding-left: 20px;
    margin-left: 0;
  }

  img {
    display: block;
    margin: 0 auto;
  }

  h3 {
    font-size: 28px;
  }
}
