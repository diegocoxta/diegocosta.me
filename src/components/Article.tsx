import React from 'react';
import styled from 'styled-components';
import { graphql } from 'gatsby';

import Title from './Title';
import TagList from './TagList';
import BodyPost from './BodyPost';

const Container = styled.article`
  margin: 20px 0 60px 0;
  font-weight: 400;
  color: ${({ theme }) => theme.textColor};
`;

const MetaAttributes = styled.p`
  color: ${({ theme }) => theme.subtitleColor};
  font-size: 19px;
  margin: 10px 0;
`;

type Nullable<T> = T | null;

export type ArticleProps = {
  readingTime?: number;
  date?: Nullable<string>;
  language?: Nullable<string>;
  content?: Nullable<string>;
  kind?: Nullable<string>;
  tags?: string[] | null;
  title: string;
  url?: string | null;
  showContent?: boolean | null;
};

export default function Article(props: ArticleProps): React.ReactElement {
  const getReadingTime = () => {
    if (!props.readingTime) {
      return undefined;
    }

    if (props.readingTime < 1) {
      return `· Less than 1 minute of reading`;
    }

    return `· ${props.readingTime?.toFixed()} minutes of reading`;
  };

  const date =
    props.date &&
    new Date(props.date).toLocaleDateString('en-us', {
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
    });

  return (
    <Container data-testid="article-item">
      <Title to={props.url ?? ''}>{props.title}</Title>
      {props.kind === 'articles' && (
        <>
          <MetaAttributes>
            {date} {getReadingTime()} · In {props.language}
          </MetaAttributes>
          <TagList tags={props.tags} />
        </>
      )}
      {props.content && <BodyPost html={props.content} />}
    </Container>
  );
}

export const query = graphql`
  fragment ArticleInformation on MarkdownRemark {
    html
    excerpt
    fields {
      collection
      slug
      readingTime {
        minutes
      }
    }
    frontmatter {
      date
      title
      description
      tags
      flags
      language
    }
  }
`;
