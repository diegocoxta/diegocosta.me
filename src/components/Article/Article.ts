import styled from 'styled-components';

export default styled.article`
  margin: 0 0 60px 0;
  font-weight: 400;
  color: ${({ theme }) => theme.textColor};
`;

export const Content = styled.section`
  line-height: 1.5;

  a {
    color: ${({ theme }) => theme.accentColor};
    text-decoration: none;
    border-bottom: 1px solid ${({ theme }) => theme.backgroundColor};

    :hover,
    :focus {
      border-bottom: 1px solid ${({ theme }) => theme.accentColor};
      outline: none;
    }
  }

  blockquote {
    border-left: 5px solid ${({ theme }) => theme.accentColor};
    padding-left: 20px;
  }

  .gatsby-highlight {
    padding: 10px 0;
  }

  pre[class*='language-'],
  code[class*='language-'],
  .gatsby-highlight {
    background-color: ${({ theme }) => theme.backgroundColor};
    font-family: monospace;
  }

  .line-numbers-rows,
  .line-numbers-rows span::before {
    border: 0.5px solid ${({ theme }) => theme.backgroundColor};
  }

  img {
    box-shadow: none !important;
  }
`;
