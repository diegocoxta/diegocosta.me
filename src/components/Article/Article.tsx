import React from 'react';
import styled from 'styled-components';
import kebabCase from 'lodash.kebabcase';
import { MDXRenderer } from 'gatsby-plugin-mdx';

import { usei18n, Link } from '~/utils/i18n';

import FixedContainer from '~/components/FixedContainer';

const Content = styled.article`
  margin: 0 0 60px 0;
  font-weight: 400;
  color: ${({ theme }) => theme.textColor};
`;

const Header = styled.header`
  margin: 20px 0 0 0;
`;

const Title = styled.h2`
  font-size: 36px;
  margin: 0;
`;

const CustomLink = styled(Link)`
  color: ${({ theme }) => theme.titleColor};
  box-shadow: none;
  text-decoration: none;

  :hover,
  :focus {
    border-bottom: 1px solid ${({ theme }) => theme.titleColor};
    outline: none;
  }
`;

const TagLink = styled(Link)`
  text-decoration: none;
  box-shadow: none;
  font-size: 14px;
  color: ${({ theme }) => theme.accentColor};
  font-weight: 700;
  text-transform: lowercase;

  :hover,
  :focus {
    border-bottom: 1px solid ${({ theme }) => theme.accentColor};
    outline: none;
  }
`;

const TagList = styled.ul`
  display: flex;
  flex-wrap: wrap;
  margin-bottom: 0;
  padding: 0;
`;

const TagItem = styled.li`
  list-style: none;
  padding: 0 10px 0 0;
  margin-bottom: 5px;
`;

const Details = styled.p`
  color: ${({ theme }) => theme.subtitleColor};
  font-size: 18px;
  margin: 5px 0;
`;

export const Body = styled.section`
  line-height: 1.5;

  a {
    color: ${({ theme }) => theme.accentColor};
    text-decoration: none;
    border-bottom: 1px solid ${({ theme }) => theme.backgroundColor};

    :hover,
    :focus {
      border-bottom: 1px solid ${({ theme }) => theme.accentColor};
      outline: none;
    }
  }

  blockquote {
    border-left: 5px solid ${({ theme }) => theme.accentColor};
    padding-left: 20px;
  }

  .gatsby-highlight {
    padding: 10px 0;
  }

  pre[class*='language-'],
  code[class*='language-'],
  .gatsby-highlight {
    background-color: ${({ theme }) => theme.backgroundColor};
    font-family: monospace;
    white-space: pre-wrap;
  }

  .line-numbers-rows,
  .line-numbers-rows span::before {
    border: 0.5px solid ${({ theme }) => theme.backgroundColor};
  }

  img {
    box-shadow: none !important;
  }

  h3 {
    font-size: 28px;
  }
`;

export interface ArticleProps {
  title: string;
  url?: string | null;
  readingTime?: number;
  language?: string | null;
  date?: string;
  tags?: string[] | null;
  description?: string | null;
  mdxContent?: string | null;
  showArticleDetails?: boolean;
}

export default function Article(props: ArticleProps): React.ReactElement {
  const i18n = usei18n();

  const pageLanguage = i18n.getCurrentLanguage();

  const getReadingTime = () => {
    if (!props.readingTime) {
      return undefined;
    }

    const lessThan1Minute = i18n.getTranslationFor('article.lessThan1Minute');
    const ofReading = i18n.getTranslationFor('article.ofReading');
    const minutes = i18n.getTranslationFor('article.minutes');

    if (props.readingTime < 1) {
      return ` · ${lessThan1Minute} ${ofReading}`;
    }

    return ` · ${props.readingTime?.toFixed()} ${minutes} ${ofReading}`;
  };

  const date =
    props.date &&
    new Date(props.date).toLocaleDateString(pageLanguage, {
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
    });

  const language = () => {
    if (!props.language) {
      return;
    }

    const languagePrefix = i18n.getTranslationFor('article.languagePrefix');
    const languageName = i18n.getTranslationFor(`languages.${props.language}`);

    return ` · ${languagePrefix} ${languageName}`;
  };

  const Tags = props.showArticleDetails && props.tags && (
    <TagList data-testid="article-header-tags">
      {props.tags.map((tag: string, index: number) => (
        <TagItem key={`${index}-${tag}`} data-testid="article-header-tag">
          <TagLink to={`/tags/${kebabCase(tag ?? '')}`}>#{kebabCase(tag ?? '')}</TagLink>
        </TagItem>
      ))}
    </TagList>
  );

  return (
    <FixedContainer>
      <Content data-testid="article-item">
        <Header>
          <Title>
            {props.url ? (
              <CustomLink to={props.url} data-testid="article-header-custom-link" language={props.language}>
                {props.title}
              </CustomLink>
            ) : (
              props.title
            )}
          </Title>
          {props.showArticleDetails && (
            <Details>
              {date} {getReadingTime()} {language()}
            </Details>
          )}
          {Tags}
        </Header>
        {props.description && <Body dangerouslySetInnerHTML={{ __html: props.description }} />}
        {props.mdxContent && (
          <Body>
            <MDXRenderer>{props.mdxContent}</MDXRenderer>
          </Body>
        )}
      </Content>
    </FixedContainer>
  );
}

Article.defaultProps = {
  showArticleDetails: true,
};
