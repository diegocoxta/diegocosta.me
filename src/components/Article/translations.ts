export type Translations = typeof en;

const en = {
  lessThan1Minute: 'Less than 1 minute',
  minutes: 'minutes',
  ofReading: 'of reading',
};

const pt: Translations = {
  lessThan1Minute: 'Menos de 1 minuto',
  minutes: 'minutos',
  ofReading: 'de leitura',
};

const translations: { [key: string]: Translations } = {
  pt,
  en,
};

export default translations;
