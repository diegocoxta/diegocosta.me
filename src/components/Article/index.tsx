import React from 'react';
import { graphql } from 'gatsby';

import Article from './Article';

interface Props {
  article?: Queries.ArticleInformationFragment | null;
  showContent?: boolean | null;
}

export default (props: Props): React.ReactElement => {
  const { article, showContent } = props;
  const { frontmatter, fields, html, excerpt } = article ?? {};

  return (
    <Article
      kind={fields?.collection}
      title={frontmatter?.title ?? ''}
      date={frontmatter?.date}
      url={fields?.slug}
      tags={frontmatter?.tags as string[]}
      readingTime={fields?.readingTime?.minutes ?? 0}
      language={fields?.language}
      content={showContent ? html : frontmatter?.description || excerpt}
    />
  );
};

export const query = graphql`
  fragment ArticleInformation on MarkdownRemark {
    html
    excerpt
    fields {
      collection
      slug
      language
      readingTime {
        minutes
      }
    }
    frontmatter {
      date
      title
      description
      tags
      homepage_view_full_article
      status
    }
  }
`;
