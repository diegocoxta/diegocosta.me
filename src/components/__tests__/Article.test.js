import React from 'react';
import { render } from '@testing-library/react';

import Article from '../Article';

describe('<Article />', () => {
  it('renders properly', () => {
    const { baseElement, getByTestId, getByText } = render(
      <Article
        title="Awesome Article"
        url="/awesome-article"
        readingTime={5}
        language="en"
        date="2020-01-18T22:12:03.284Z"
        tags={['jest', 'testing-library']}
        kind="articles"
      />
    );
    expect(getByText('Awesome Article')).toBeTruthy();
    expect(getByTestId('title').href).toBe('http://localhost/awesome-article');
    expect(baseElement).toMatchSnapshot();
  });

  it('renders properly the body content', () => {
    const { baseElement, getByText } = render(
      <Article
        title="Awesome Article"
        url="/awesome-article"
        readingTime={5}
        language="pt"
        date="2020-01-18T22:12:03.284Z"
        content="Article as a Children"
      />
    );

    expect(getByText('Article as a Children')).toBeTruthy();
    expect(baseElement).toMatchSnapshot();
  });

  it('renders properly the description', () => {
    const { baseElement, getByText } = render(
      <Article
        title="Awesome Article"
        url="/awesome-article"
        readingTime={5}
        language="pt"
        date="2020-01-18T22:12:03.284Z"
        content="Article as a Children"
      />
    );

    expect(getByText('Article as a Children')).toBeTruthy();
    expect(baseElement).toMatchSnapshot();
  });
});
