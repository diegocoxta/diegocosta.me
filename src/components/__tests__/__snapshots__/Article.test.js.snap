// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Article /> renders properly 1`] = `
.c1 {
  box-shadow: none;
  text-decoration: none;
  font-size: 36px;
  margin: 0;
  font-weight: 700;
}

.c1:hover,
.c1:focus {
  border-bottom: 1px solid undefined;
  outline: none;
}

.c3 {
  display: flex;
  flex-wrap: wrap;
  padding: 0;
  margin: 0;
}

.c4 {
  list-style: none;
  padding: 0 10px 0 0;
}

.c5 {
  text-decoration: none;
  box-shadow: none;
  font-size: 16px;
  font-weight: 700;
  text-transform: lowercase;
}

.c5:hover,
.c5:focus {
  border-bottom: 1px solid;
  outline: none;
}

.c0 {
  margin: 20px 0 60px 0;
  font-weight: 400;
}

.c2 {
  font-size: 19px;
  margin: 10px 0;
}

<body>
  <div>
    <article
      class="c0"
      data-testid="article-item"
    >
      <a
        class="c1"
        data-testid="title"
        href="/awesome-article"
      >
        Awesome Article
      </a>
      <p
        class="c2"
      >
        01/18/2020
         
        · 5 minutes of reading
         · In 
        en
      </p>
      <ul
        class="c3"
        data-testid="taglist-list"
      >
        <li
          class="c4"
          data-testid="taglist-item"
        >
          <a
            class="c5"
            href="/tags/jest"
          >
            #jest
          </a>
        </li>
        <li
          class="c4"
          data-testid="taglist-item"
        >
          <a
            class="c5"
            href="/tags/testing-library"
          >
            #testing-library
          </a>
        </li>
      </ul>
    </article>
  </div>
</body>
`;

exports[`<Article /> renders properly the body content 1`] = `
.c1 {
  box-shadow: none;
  text-decoration: none;
  font-size: 36px;
  margin: 0;
  font-weight: 700;
}

.c1:hover,
.c1:focus {
  border-bottom: 1px solid undefined;
  outline: none;
}

.c2 {
  line-height: 1.5;
  font-size: 19px;
  margin-top: 10px;
}

.c2 a {
  text-decoration: none;
  border-bottom: 1px solid;
}

.c2 a:hover,
.c2 a:focus {
  border-bottom: 1px solid;
  outline: none;
}

.c2 code {
  background: 40;
  padding: 2px 6px;
  border-radius: 10px;
  font-size: 16px;
}

.c2 blockquote {
  border-left: 5px solid;
  padding-left: 20px;
  margin-left: 0;
}

.c2 img {
  box-shadow: none!important;
}

.c2 h3 {
  font-size: 28px;
}

.c2 table.uses-table {
  width: 100%;
  margin: 40px 0;
}

.c2 table.uses-table td {
  text-align: center;
}

.c0 {
  margin: 20px 0 60px 0;
  font-weight: 400;
}

<body>
  <div>
    <article
      class="c0"
      data-testid="article-item"
    >
      <a
        class="c1"
        data-testid="title"
        href="/awesome-article"
      >
        Awesome Article
      </a>
      <section
        class="c2"
        html="Article as a Children"
      >
        Article as a Children
      </section>
    </article>
  </div>
</body>
`;

exports[`<Article /> renders properly the description 1`] = `
.c1 {
  box-shadow: none;
  text-decoration: none;
  font-size: 36px;
  margin: 0;
  font-weight: 700;
}

.c1:hover,
.c1:focus {
  border-bottom: 1px solid undefined;
  outline: none;
}

.c2 {
  line-height: 1.5;
  font-size: 19px;
  margin-top: 10px;
}

.c2 a {
  text-decoration: none;
  border-bottom: 1px solid;
}

.c2 a:hover,
.c2 a:focus {
  border-bottom: 1px solid;
  outline: none;
}

.c2 code {
  background: 40;
  padding: 2px 6px;
  border-radius: 10px;
  font-size: 16px;
}

.c2 blockquote {
  border-left: 5px solid;
  padding-left: 20px;
  margin-left: 0;
}

.c2 img {
  box-shadow: none!important;
}

.c2 h3 {
  font-size: 28px;
}

.c2 table.uses-table {
  width: 100%;
  margin: 40px 0;
}

.c2 table.uses-table td {
  text-align: center;
}

.c0 {
  margin: 20px 0 60px 0;
  font-weight: 400;
}

<body>
  <div>
    <article
      class="c0"
      data-testid="article-item"
    >
      <a
        class="c1"
        data-testid="title"
        href="/awesome-article"
      >
        Awesome Article
      </a>
      <section
        class="c2"
        html="Article as a Children"
      >
        Article as a Children
      </section>
    </article>
  </div>
</body>
`;
